File: unreal/OracleStrategyAdapter.cpp

#include "OracleStrategyAdapter.h"
#include "Regex.h"
#include "JsonObjectConverter.h"

void UOracleStrategyAdapter::InitDefaults(){
  Turn = 0; Provinces.Empty(); Knobs.Empty(); Rules.Empty(); Events.Empty(); SnapshotsByTurn.Empty();
  FProvinceStats Market; Market.Metrics.Add("security",42); Market.Metrics.Add("prosperity",55); Market.Metrics.Add("control",50);
  Provinces.Add("Market", Market);
  FProvinceStats Harbor; Harbor.Metrics.Add("security",37); Harbor.Metrics.Add("prosperity",48); Harbor.Metrics.Add("control",52);
  Provinces.Add("Harbor", Harbor);
  Knobs.Add("area_half_life", 720);
  SnapshotsByTurn.Add(0, AutosaveJson());
}

void UOracleStrategyAdapter::NextTurn(){
  Turn += 1;
  SnapshotsByTurn.Add(Turn, AutosaveJson());
}

void UOracleStrategyAdapter::AreaBump(const FString& Province, const FString& Dim, float Delta){
  FProvinceStats* P = Provinces.Find(Province);
  if(!P){ FProvinceStats NewP; Provinces.Add(Province, NewP); P = Provinces.Find(Province); }
  float* V = P->Metrics.Find(Dim);
  if(!V){ P->Metrics.Add(Dim, 50.f); V = P->Metrics.Find(Dim); }
  *V += Delta;
  FTurnEvent E; E.Turn = Turn; E.Verb = "AreaBump"; E.Ctx.Add("province", Province); E.Ctx.Add("dim", Dim); E.Ctx.Add("delta", FString::SanitizeFloat(Delta));
  Events.Add(E);
}

void UOracleStrategyAdapter::DslLoad(const FString& Name, const FString& Rule){
  Rules.Add(Name, Rule);
  FTurnEvent E; E.Turn = Turn; E.Verb = "DSLLoaded"; E.Ctx.Add("name", Name); E.Ctx.Add("size", FString::FromInt(Rule.Len())); Events.Add(E);
}

bool UOracleStrategyAdapter::DslExec(const FString& Name, const FString& Province, const FString& Outcome, TArray<FString>& AppliedDims, TArray<float>& AppliedDeltas){
  AppliedDims.Empty(); AppliedDeltas.Empty();
  const FString* Txt = Rules.Find(Name);
  if(!Txt) return false;
  if(!(Txt->Contains("OUTCOMES") && Txt->Contains("area."))) return false;

  TArray<FString> Lines; Txt->ParseIntoArrayLines(Lines, true);
  bool bPick = false;
  FRegexPattern Pat(TEXT("area\\.(\\w+)\\s*([+\\-]=)\\s*([\\-0-9\\.]+)"));

  for (const FString& L : Lines){
    FString T = L; T.TrimStartAndEndInline();
    FRegexPattern OutPat(FString::Printf(TEXT("^%s\\s*:"), *Outcome));
    if (FRegexMatcher(OutPat, T).FindNext()){ bPick = true; continue; }
    if (bPick && FRegexMatcher(FRegexPattern(TEXT("^\\w+\\s*:")), T).FindNext()) break;

    if (bPick){
      FRegexMatcher M(Pat, T);
      if (M.FindNext()){
        FString Dim = M.GetCaptureGroup(1);
        FString Op  = M.GetCaptureGroup(2);
        float Num   = FCString::Atof(*M.GetCaptureGroup(3));
        float Delta = (Op == "+=") ? Num : -Num;
        AreaBump(Province, Dim, Delta);
        AppliedDims.Add(Dim); AppliedDeltas.Add(Delta);
      }
    }
  }
  FTurnEvent Ev; Ev.Turn = Turn; Ev.Verb = "EncounterCommitted"; Ev.Ctx.Add("province", Province); Ev.Ctx.Add("rule", Name); Ev.Ctx.Add("outcome", Outcome);
  Events.Add(Ev);
  return true;
}

FString UOracleStrategyAdapter::AutosaveJson() const{
  FStrategySave Save; Save.Turn = Turn; Save.Provinces = Provinces; Save.Knobs = Knobs; Save.Events = Events;
  FString OutJson; FJsonObjectConverter::UStructToJsonObjectString(Save, OutJson);
  return OutJson;
}

void UOracleStrategyAdapter::LoadJson(const FString& Json){
  FStrategySave Save;
  if (FJsonObjectConverter::JsonObjectStringToUStruct(Json, &Save, 0, 0)){
    Turn = Save.Turn; Provinces = Save.Provinces; Knobs = Save.Knobs; Events = Save.Events;
    SnapshotsByTurn.Add(Turn, Json);
  }
}

void UOracleStrategyAdapter::UndoToTurn(int32 TargetTurn){
  if (const FString* Json = SnapshotsByTurn.Find(TargetTurn)){
    LoadJson(*Json);
  }
}

File: unity/OracleStrategyAdapter.cs

using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using UnityEngine;

[Serializable] public class ProvinceStats : SerializableDictionary<string, float> { }
[Serializable] public class Provinces : SerializableDictionary<string, ProvinceStats> { }
[Serializable] public class TurnEvent { public int turn; public string verb; public Dictionary<string, object> ctx = new(); }

[Serializable]
public class StrategySave {
    public int turn;
    public Provinces provinces = new();
    public SerializableDictionary<string,float> knobs = new();
    public List<TurnEvent> events = new();
}

public class OracleStrategyAdapter {
    public int Turn { get; private set; } = 0;
    public Provinces Provinces = new();
    public Dictionary<string,float> Knobs = new() { {"area_half_life", 720} };
    private readonly Dictionary<string,string> _rules = new();
    private readonly List<TurnEvent> _events = new();
    private readonly Dictionary<int, StrategySave> _snapshots = new();

    public OracleStrategyAdapter() {
        Provinces["Market"] = new ProvinceStats {["security"]=42,["prosperity"]=55,["control"]=50};
        Provinces["Harbor"] = new ProvinceStats {["security"]=37,["prosperity"]=48,["control"]=52};
        _snapshots[0] = MakeSave();
    }

    private void Log(string verb, Dictionary<string, object> ctx){ _events.Add(new TurnEvent { turn = Turn, verb = verb, ctx = ctx }); }

    public void NextTurn(){ Turn += 1; _snapshots[Turn] = MakeSave(); }

    public void AreaBump(string province, string dim, float delta){
        if (!Provinces.ContainsKey(province)) Provinces[province] = new ProvinceStats();
        if (!Provinces[province].ContainsKey(dim)) Provinces[province][dim] = 50;
        Provinces[province][dim] += delta;
        Log("AreaBump", new(){["province"]=province,["dim"]=dim,["delta"]=delta});
    }

    public void DslLoad(string name, string rule){ _rules[name] = rule ?? ""; Log("DSLLoaded", new(){["name"]=name,"size"]=_rules[name].Length}); }

    public (bool matched, List<(string dim, float delta)> applied) DslExec(string name, string province, string outcome){
        if (!_rules.TryGetValue(name, out var txt)) return (false, new());
        bool ok = txt.Contains("OUTCOMES") && txt.Contains("area.");
        var applied = new List<(string,float)>();
        if (ok){
            var lines = txt.Split(new[]{'\n','\r'}, StringSplitOptions.RemoveEmptyEntries);
            bool pick=false;
            var pat = new Regex(@"area\.(\w+)\s*([+\-]=)\s*([\-0-9\.]+)");
            foreach(var ln in lines){
                var trimmed = ln.Trim();
                if (Regex.IsMatch(trimmed, @"^"+Regex.Escape(outcome)+@"\s*:")) { pick=true; continue; }
                if (pick && Regex.IsMatch(trimmed, @"^\w+\s*:")) break;
                if (pick){
                    var m = pat.Match(trimmed);
                    if (m.Success){
                        string dim = m.Groups[1].Value;
                        string op  = m.Groups[2].Value;
                        float num  = float.Parse(m.Groups[3].Value, System.Globalization.CultureInfo.InvariantCulture);
                        float delta = (op == "+=") ? num : -num;
                        AreaBump(province, dim, delta);
                        applied.Add((dim, delta));
                    }
                }
            }
            Log("EncounterCommitted", new(){["province"]=province,["rule"]=name,["outcome"]=outcome,["applied"]=applied.Count});
        }
        return (ok, applied);
    }

    public StrategySave MakeSave(){
        return new StrategySave {
            turn = Turn, provinces = Provinces, knobs = new SerializableDictionary<string,float>(Knobs), events = new List<TurnEvent>(_events)
        };
    }
    public string AutosaveJson(){ return JsonUtility.ToJson(MakeSave(), true); }
    public void LoadJson(string json){
        var s = JsonUtility.FromJson<StrategySave>(json);
        Turn = s.turn; Provinces = s.provinces; Knobs = new(s.knobs); _events.Clear(); _events.AddRange(s.events);
        _snapshots[Turn] = s;
    }
    public void UndoToTurn(int t){
        if (!_snapshots.TryGetValue(t, out var snap)) throw new Exception($"No snapshot for turn {t}");
        Turn = snap.turn; Provinces = snap.provinces; Knobs = new(snap.knobs);
        Log("UndoToTurn", new(){["turn"]=t});
    }
}
